File: frontend/services/mcp.ts
import { get, post, put, del } from './api';
import { ApiResponse } from '@/types'; // Corrected: Import ApiResponse from its source
import { AppError } from '@/utils/errorHandling';

// --- TypeScript Interfaces (mirroring Pydantic schemas) ---

export interface MCPServerConfigBase {
  name: string;
  command: 'docker'; // Enforce 'docker' as per backend validation
  args: string[];
  env?: Record<string, string>;
  enabled: boolean;
}

export interface MCPServerConfigCreate extends MCPServerConfigBase {}

export interface MCPServerConfigUpdate {
  name?: string;
  args?: string[];
  env?: Record<string, string>;
  enabled?: boolean;
}

export interface MCPServerConfigResponse extends MCPServerConfigBase {
  id: string;
  user_id: string;
  created_at: string; // Representing datetime as string
  updated_at: string; // Representing datetime as string
}

export interface MCPServerStatus {
  id: string;
  name: string;
  enabled: boolean;
  status: 'running' | 'stopped' | 'error'; // Literal types for status
  container_id?: string;
  error_message?: string;
}

export interface MCPConfigJSON {
  mcpServers: Record<string, Record<string, any>>; // Using 'any' for flexibility as per backend schema
}

// --- Helper function to handle API response ---
const handleResponse = <T>(response: ApiResponse<T>, errorMessage: string): T => {
  if (response.error) {
    // Re-throw the specific error object if available, otherwise a generic one
    throw response.errorObject || new Error(response.error || errorMessage);
  }
  if (response.data === undefined) {
    // Handle cases where data might be unexpectedly undefined even without an error
    // For DELETE (204 No Content), undefined data is expected and okay.
    // For other methods, this might indicate an issue.
    // We return undefined here, callers might need to check.
    // Or throw new Error('API returned undefined data unexpectedly.');
  }
  // Type assertion needed because T could be void for DELETE
  return response.data as T;
};


// --- API Service Functions ---

const MCP_BASE_URL = '/mcp/configs'; // Base path for MCP config endpoints

/**
 * Fetches all MCP server configurations for the current user.
 */
export const getMcpConfigs = async (): Promise<MCPServerConfigResponse[]> => {
  const response = await get<MCPServerConfigResponse[]>(MCP_BASE_URL);
  return handleResponse(response, 'Failed to fetch MCP configurations');
};

/**
 * Fetches a specific MCP server configuration by its ID.
 * @param configId - The ID of the configuration to fetch.
 */
export const getMcpConfig = async (configId: string): Promise<MCPServerConfigResponse> => {
  const response = await get<MCPServerConfigResponse>(`${MCP_BASE_URL}/${configId}`);
  return handleResponse(response, `Failed to fetch MCP configuration ${configId}`);
};

/**
 * Creates a new MCP server configuration. (Admin only)
 * @param configData - The data for the new configuration.
 */
export const createMcpConfig = async (configData: MCPServerConfigCreate): Promise<MCPServerConfigResponse> => {
  const response = await post<MCPServerConfigResponse>(MCP_BASE_URL, configData);
  return handleResponse(response, 'Failed to create MCP configuration');
};

/**
 * Updates an existing MCP server configuration. (Admin only)
 * @param configId - The ID of the configuration to update.
 * @param configData - The update data.
 */
export const updateMcpConfig = async (configId: string, configData: MCPServerConfigUpdate): Promise<MCPServerConfigResponse> => {
  const response = await put<MCPServerConfigResponse>(`${MCP_BASE_URL}/${configId}`, configData);
  return handleResponse(response, `Failed to update MCP configuration ${configId}`);
};

/**
 * Deletes an MCP server configuration. (Admin only)
 * @param configId - The ID of the configuration to delete.
 */
export const deleteMcpConfig = async (configId: string): Promise<void> => {
  // DELETE requests often return 204 No Content, so data might be undefined
  const response = await del<void>(`${MCP_BASE_URL}/${configId}`);
  // Use handleResponse, but expect undefined data for success
  handleResponse(response, `Failed to delete MCP configuration ${configId}`);
};

/**
 * Fetches the status of a specific MCP server.
 * @param configId - The ID of the configuration whose status to fetch.
 */
export const getMcpConfigStatus = async (configId: string): Promise<MCPServerStatus> => {
  const response = await get<MCPServerStatus>(`${MCP_BASE_URL}/${configId}/status`);
  return handleResponse(response, `Failed to fetch status for MCP configuration ${configId}`);
};

/**
 * Starts a specific MCP server. (Admin only)
 * @param configId - The ID of the configuration to start.
 */
export const startMcpServer = async (configId: string): Promise<MCPServerStatus> => {
  const response = await post<MCPServerStatus>(`${MCP_BASE_URL}/${configId}/start`);
  return handleResponse(response, `Failed to start MCP server ${configId}`);
};

/**
 * Stops a specific MCP server. (Admin only)
 * @param configId - The ID of the configuration to stop.
 */
export const stopMcpServer = async (configId: string): Promise<MCPServerStatus> => {
  const response = await post<MCPServerStatus>(`${MCP_BASE_URL}/${configId}/stop`);
  return handleResponse(response, `Failed to stop MCP server ${configId}`);
};

/**
 * Restarts a specific MCP server. (Admin only)
 * @param configId - The ID of the configuration to restart.
 */
export const restartMcpServer = async (configId: string): Promise<MCPServerStatus> => {
  const response = await post<MCPServerStatus>(`${MCP_BASE_URL}/${configId}/restart`);
  return handleResponse(response, `Failed to restart MCP server ${configId}`);
};

/**
 * Fetches the MCP configuration in JSON format (e.g., for Claude Desktop).
 */
export const getMcpConfigJson = async (): Promise<MCPConfigJSON> => {
  const response = await get<MCPConfigJSON>(`${MCP_BASE_URL}/json`);
  return handleResponse(response, 'Failed to fetch MCP configuration JSON');
};

File: backend/app/api/routes/mcp.py
"""
MCP (Model Context Protocol) API routes.

This module defines the API endpoints for managing MCP server configurations
and controlling MCP servers.
"""

from typing import List
# Import status with an alias to avoid conflicts
from fastapi import APIRouter, Depends, HTTPException, status as fastapi_status
from sqlalchemy.orm import Session

from app.db.base import get_db
from app.models.user import User, UserRole
from app.schemas.mcp import (
    MCPServerConfigCreate,
    MCPServerConfigUpdate,
    MCPServerConfigResponse,
    MCPServerStatus,
    MCPConfigJSON
)
# Import functions directly from the new package
from app.services.mcp_config_service import (
    create_config,
    get_configs_by_user,
    get_config_by_id,
    update_config,
    delete_config,
    stop_server,
    get_config_status,
    start_server,
    restart_server,
    generate_mcp_config_json
)
from app.utils.deps import get_current_user

router = APIRouter()

@router.post("/configs", response_model=MCPServerConfigResponse)
async def create_mcp_config(
    config: MCPServerConfigCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Create a new MCP server configuration.

    Only admin users can create MCP configurations.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can create MCP configurations"
        )

    db_config = create_config(db, config, current_user.id) # Use imported function
    return db_config

@router.get("/configs", response_model=List[MCPServerConfigResponse])
async def get_mcp_configs(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get all MCP server configurations for the current user.
    """
    return get_configs_by_user(db, current_user.id) # Use imported function

@router.get("/configs/{config_id}", response_model=MCPServerConfigResponse)
async def get_mcp_config(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get an MCP server configuration by ID.
    """
    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )
    return db_config

@router.put("/configs/{config_id}", response_model=MCPServerConfigResponse)
async def update_mcp_config(
    config_id: str,
    config_update: MCPServerConfigUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Update an MCP server configuration.

    Only admin users can update MCP configurations.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can update MCP configurations"
        )

    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    updated_config = update_config(db, config_id, config_update) # Use imported function
    return updated_config

@router.delete("/configs/{config_id}", status_code=fastapi_status.HTTP_204_NO_CONTENT) # Use alias
async def delete_mcp_config(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Delete an MCP server configuration.

    Only admin users can delete MCP configurations.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can delete MCP configurations"
        )

    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    # Stop the server if it's running
    try:
        stop_server(db, config_id) # Use imported function
    except Exception as e:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail=f"Failed to stop MCP server: {str(e)}"
        )

    # Delete the configuration
    success = delete_config(db, config_id) # Use imported function
    if not success:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail="Failed to delete MCP configuration"
        )

@router.get("/configs/{config_id}/status", response_model=MCPServerStatus)
async def get_mcp_config_status(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get the status of an MCP server.
    """
    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    status_result = get_config_status(db, config_id) # Use imported function
    if not status_result:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail="Failed to get MCP server status"
        )

    return status_result

@router.post("/configs/{config_id}/start", response_model=MCPServerStatus)
async def start_mcp_server(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Start an MCP server.

    Only admin users can start MCP servers.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can start MCP servers"
        )

    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    if not db_config.config or not db_config.config.get('enabled', False):
        raise HTTPException(
            status_code=fastapi_status.HTTP_400_BAD_REQUEST, # Use alias
            detail="Cannot start disabled MCP server"
        )

    status_result = start_server(db, config_id) # Use imported function
    if not status_result:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail="Failed to start MCP server"
        )

    if status_result.status == "error":
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail=f"Failed to start MCP server: {status_result.error_message}"
        )

    return status_result

@router.post("/configs/{config_id}/stop", response_model=MCPServerStatus)
async def stop_mcp_server(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Stop an MCP server.

    Only admin users can stop MCP servers.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can stop MCP servers"
        )

    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    status_result = MCPConfigService.stop_server(db, config_id) # Renamed variable
    if not status_result:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail="Failed to stop MCP server"
        )

    if status_result.status == "error":
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail=f"Failed to stop MCP server: {status_result.error_message}"
        )

    return status_result

@router.post("/configs/{config_id}/restart", response_model=MCPServerStatus)
async def restart_mcp_server(
    config_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Restart an MCP server.

    Only admin users can restart MCP servers.
    """
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(
            status_code=fastapi_status.HTTP_403_FORBIDDEN, # Use alias
            detail="Only admin users can restart MCP servers"
        )

    db_config = get_config_by_id(db, config_id) # Use imported function
    if not db_config or db_config.user_id != current_user.id:
        raise HTTPException(
            status_code=fastapi_status.HTTP_404_NOT_FOUND, # Use alias
            detail="MCP configuration not found"
        )

    status_result = restart_server(db, config_id) # Use imported function
    if not status_result:
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail="Failed to restart MCP server"
        )

    if status_result.status == "error":
        raise HTTPException(
            status_code=fastapi_status.HTTP_500_INTERNAL_SERVER_ERROR, # Use alias
            detail=f"Failed to restart MCP server: {status_result.error_message}"
        )

    return status_result

@router.get("/configs/export/json", response_model=MCPConfigJSON) # Changed path
async def get_mcp_config_json(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get the MCP configuration JSON for Claude Desktop.
    """
    return generate_mcp_config_json(db, current_user.id) # Use imported function
